grpc-client-server/prober_server/main.go

```diff
@@ -4,44 +4,141 @@ import (
+	port         = flag.Int("port", 50051, "The server port")
+	LatencyGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "latency_gauge",
+		Help: "The latency of the requests to the endpoint",
+	}, []string{"endpoint", "type"})
+	timeSince = time.Since
 )
 
+type ResponseTime struct {
+	Ttfb time.Duration
+	Ttlb time.Duration
+}
+
 // server is used to implement prober.ProberServer.
-type server struct {
+type Server struct {
 	pb.UnimplementedProberServer
 }
 
-func (s *server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
-	// TODO: support a number of repetitions and return average latency
-	start := time.Now()
-	_, _ = http.Get(in.GetEndpoint())	// TODO: add error handling here and check the response code
-	elapsed := time.Since(start)
-	elapsedMsecs := float32(elapsed / time.Millisecond)
+func (s *Server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
+	ttfbTotal := time.Duration(0)
+	ttlbTotal := time.Duration(0)
+	failed := 0
+	latestTtfb := time.Duration(0)
+
+	c := &http.Client{
+		Transport: &TimedRoundTripper{
+			defaultTripper: http.DefaultTransport,
+			recordTime: func(t time.Duration) {
+				latestTtfb = t
+				ttfbTotal += t
+			},
+		},
+	}
+
+	for i := 0; i < int(in.GetNumberOfRequests()); i++ {
+		ttlb, err := TimedProbe(c, in.GetEndpoint())
+		if err != nil {
+			log.Printf("could not probe: %v", err)
+			failed++
+			continue
+		}
+		LatencyGauge.WithLabelValues(in.GetEndpoint(), "TTFB").Set(float64(latestTtfb / time.Millisecond))
+		LatencyGauge.WithLabelValues(in.GetEndpoint(), "TTLB").Set(float64(ttlb / time.Millisecond))
+		ttlbTotal += ttlb
+	}
+	ttfbAverage := time.Duration(float32(ttfbTotal) / float32(in.NumberOfRequests-int32(failed)))
+	ttlbAverage := time.Duration(float32(ttlbTotal) / float32(in.NumberOfRequests-int32(failed)))
 
-	return &pb.ProbeReply{Result: elapsedMsecs}, nil
+	return &pb.ProbeReply{
+		TtfbAverageResponseTime: durationpb.New(ttfbAverage),
+		TtlbAverageResponseTime: durationpb.New(ttlbAverage),
+		FailedRequests:          int32(failed),
+	}, nil
+}
+func InitMonitoring() {
+	prometheus.MustRegister(LatencyGauge)
+	go func() {
+		http.Handle("/metrics", promhttp.Handler())
+		http.ListenAndServe(":2112", nil)
+	}()
 }
 
 func main() {
 	flag.Parse()
+	InitMonitoring()
 	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *port))
 	if err != nil {
 		log.Fatalf("failed to listen: %v", err)
 	}
 	s := grpc.NewServer()
-	pb.RegisterProberServer(s, &server{})
+	pb.RegisterProberServer(s, &Server{})
 	log.Printf("server listening at %v", lis.Addr())
 	if err := s.Serve(lis); err != nil {
 		log.Fatalf("failed to serve: %v", err)
 	}
 }
+
+func TimedProbe(c *http.Client, url string) (ttlb time.Duration, err error) {
+	var start time.Time
+	nullTime := time.Duration(0)
+
+	req, err := http.NewRequest("GET", url, nil)
+
+	if err != nil {
+		return nullTime, err
+	}
+	start = time.Now()
+
+	resp, err := c.Transport.RoundTrip(req)
```

Comment:
Even though you're setting up a `RoundTripper`, you can still just call `c.Do` or `c.Get` like normal - it will internally use the `RoundTripper` you set up, but as a caller, you don't need to make your code more convoluted to use it.

Reply:
Hadn't realized that, thanks!